crates/transaction-pool/src/traits.rs

```diff
@@ -272,6 +275,118 @@ pub trait PoolTransaction: fmt::Debug + Send + Sync + FromRecoveredTransaction {
+    }
+
+    /// Returns the nonce for this transaction.
+    fn nonce(&self) -> u64 {
+        self.transaction.nonce()
+    }
+
+    /// Calculates the cost that this transaction is allowed to consume:
+    ///
+    /// For EIP-1559 transactions that is `feeCap x gasLimit + transferred_value`
+    fn cost(&self) -> U256 {
+        self.cost
+    }
+
+    /// Returns the effective gas price for this transaction.
+    ///
+    /// This is `priority + basefee`for EIP-1559 and `gasPrice` for legacy transactions.
+    fn effective_gas_price(&self) -> U256 {
+        self.effective_gas_price
+    }
+
+    /// Amount of gas that should be used in executing this transaction. This is paid up-front.
+    fn gas_limit(&self) -> u64 {
+        self.transaction.gas_limit()
+    }
+
+    /// Returns the EIP-1559 Max base fee the caller is willing to pay.
+    ///
+    /// This will return `None` for non-EIP1559 transactions
+    fn max_fee_per_gas(&self) -> Option<U256> {
+        match &self.transaction.transaction {
+            Transaction::Legacy(_) => None,
+            Transaction::Eip2930(_) => None,
+            Transaction::Eip1559(tx) => Some(U256::from(tx.max_fee_per_gas)),
+        }
+    }
+
+    /// Returns the EIP-1559 Priority fee the caller is paying to the block author.
+    ///
+    /// This will return `None` for non-EIP1559 transactions
+    fn max_priority_fee_per_gas(&self) -> Option<U256> {
+        match &self.transaction.transaction {
+            Transaction::Legacy(_) => None,
+            Transaction::Eip2930(_) => None,
+            Transaction::Eip1559(tx) => Some(U256::from(tx.max_priority_fee_per_gas)),
+        }
+    }
+
+    /// Returns a measurement of the heap usage of this type and all its internals.
+    fn size(&self) -> usize {
+        self.transaction.transaction.input().len()
+    }
+}
+
+impl FromRecoveredTransaction for PooledTransaction {
+    fn from_recovered_transaction(tx: TransactionSignedEcRecovered) -> Self {
+        let size = tx.transaction.input().len();
+
+        match &tx.transaction {
+            Transaction::Legacy(t) => PooledTransaction {
+                transaction: tx.clone(),
+                cost: U256::from(t.gas_price) * U256::from(t.gas_limit) + U256::from(t.value),
+                effective_gas_price: U256::from(t.gas_price),
+                size,
+            },
+            Transaction::Eip2930(t) => PooledTransaction {
+                transaction: tx.clone(),
+                cost: U256::from(t.gas_price) * U256::from(t.gas_limit) + U256::from(t.value),
+                effective_gas_price: U256::from(t.gas_price),
+                size,
+            },
+            Transaction::Eip1559(t) => PooledTransaction {
+                transaction: tx.clone(),
+                cost: U256::from(t.max_fee_per_gas) * U256::from(t.gas_limit) + U256::from(t.value),
+                effective_gas_price: U256::from(t.max_fee_per_gas),
+                size,
+            },
+        }
```

Comment:
this entire section can be simplified, because `PooledTransaction` just wraps `TransactionSignedEcRecovered`

the only thing we need to do here is setting the other fields, like cost.

Reply:
I've removed the size field. 

Does this work for **FromRecoveredTransaction**

```rust
let mut cost =  U256::from(tx.max_fee_per_gas()) * U256::from(tx.gas_limit()) + U256::from(*tx.value());
let mut effective_gas_price = U256::from(tx.max_fee_per_gas());

match &tx.transaction {
    Transaction::Legacy(t) => {
        cost = U256::from(t.gas_price) * U256::from(t.gas_limit) + U256::from(t.value);
        effective_gas_price = U256::from(t.gas_price);
    }
    Transaction::Eip2930(t) => {
        cost = U256::from(t.gas_price) * U256::from(t.gas_limit) + U256::from(t.value);
        effective_gas_price = U256::from(t.gas_price);
    }
    _ => {}
};

PooledTransaction { transaction: tx, cost, effective_gas_price }
```

Reply:
this does a redundant calc for cost/effect, move the initializer in the other match arm

Reply:
When you say the calc for cost/effect is redundant, are you referring to calculating cost/effect on each match arm, I'm doing that because the **gas price** field is only available on the **Transaction** struct, not on **TransactionSignedEcRecovered** or **TransactionSigned**. If I move the PooledTransaction initializer inside the match, I'd have to do that for each match, similar to the initial implementation.

Reply:
this should work?

```rust


let (cost, effective_gas_price) = match &tx.transaction {
    Transaction::Legacy(t) => {
        cost = U256::from(t.gas_price) * U256::from(t.gas_limit) + U256::from(t.value);
        effective_gas_price = U256::from(t.gas_price);
    }
    Transaction::Eip2930(t) => {
        cost = U256::from(t.gas_price) * U256::from(t.gas_limit) + U256::from(t.value);
        effective_gas_price = U256::from(t.gas_price);
    }
    Eip155(t) => {
     let cost =  U256::from(tx.max_fee_per_gas()) * U256::from(tx.gas_limit()) + U256::from(*tx.value());
let effective_gas_price = U256::from(tx.max_fee_per_gas());
    }
};

```

Reply:
yeah, it would.